server/src/main/java/org/elasticsearch/action/admin/indices/settings/get/TransportGetSettingsAction.java

```diff
@@ -69,25 +73,39 @@ protected GetSettingsResponse newResponse() {
         return new GetSettingsResponse();
     }
 
+    private static boolean isFilteredRequest(GetSettingsRequest request) {
+        return CollectionUtils.isEmpty(request.names()) == false;
+    }
+
     @Override
     protected void masterOperation(GetSettingsRequest request, ClusterState state, ActionListener<GetSettingsResponse> listener) {
         Index[] concreteIndices = indexNameExpressionResolver.concreteIndices(state, request);
         ImmutableOpenMap.Builder<String, Settings> indexToSettingsBuilder = ImmutableOpenMap.builder();
+        ImmutableOpenMap.Builder<String, Settings> indexToDefaultSettingsBuilder = ImmutableOpenMap.builder();
         for (Index concreteIndex : concreteIndices) {
             IndexMetaData indexMetaData = state.getMetaData().index(concreteIndex);
             if (indexMetaData == null) {
                 continue;
             }
 
-            Settings settings = settingsFilter.filter(indexMetaData.getSettings());
+            Settings indexSettings = settingsFilter.filter(indexMetaData.getSettings());
             if (request.humanReadable()) {
-                settings = IndexMetaData.addHumanReadableSettings(settings);
+                indexSettings = IndexMetaData.addHumanReadableSettings(indexSettings);
             }
-            if (CollectionUtils.isEmpty(request.names()) == false) {
-                settings = settings.filter(k -> Regex.simpleMatch(request.names(), k));
+
+            if (isFilteredRequest(request)) {
+                indexSettings = indexSettings.filter(k -> Regex.simpleMatch(request.names(), k));
+            }
+
+            indexToSettingsBuilder.put(concreteIndex.getName(), indexSettings);
+            if (request.includeDefaults()) {
+                Settings defaultSettings = settingsFilter.filter(indexScopedSettings.diff(indexSettings, this.settings));
+                if (isFilteredRequest(request)) {
+                    defaultSettings = defaultSettings.filter(k -> Regex.simpleMatch(request.names(), k));
```

Comment:
I think that this fixes #29539 , I opened it to track that we were not filtering default settings prior to this PR.